### PR TITLE
RFC FS-1009 - optionally declare mutually referential types and modules over larger scopes within files

### PR DESCRIPTION
**Pull request by dsyme**
This is the PR for [RFC FS-1009 - Allow mutually referential types and modules over larger scopes within files](https://github.com/fsharp/FSharpLangDesign/blob/master/RFCs/FS-1009-mutually-referential-types-and-modules-single-scope.md), being pushed partly for automated testing.

Please only discuss the implementation in this PR. Discuss the RFC on [the RFC discussion thread](https://github.com/fsharp/FSharpLangDesign/issues/76).

### CODE REVIEW

File: vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 #nowarn "40"
 
-namespace Microsoft.VisualStudio.FSharp.ProjectSystem 
+namespace rec Microsoft.VisualStudio.FSharp.ProjectSystem
```
Comment (smoothdeveloper) [inline vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: I like how this is being dogfooded right in the same checkin, without the need to even have last known good carrying the feature :+1:

**Comment by dsyme** [comment]
There's one failing test at the moment

> [00:35:30] *** The following necessary lines were never matched:
> [00:35:30] ***	\(16,21-16,22\):.+error FS0430:.+The member '\( \+ \)' is used in an invalid way\. A use of '\( \+ \)' has been inferred prior to its definition.+

**Comment by dsyme** [comment]
I've updated this further and started to add a lot of testing (by taking existing tests - especially for object-oriented constructs - and adding "rec")

**Comment by dsyme** [comment]
I just updated this PR. There are a whole lot of test failures - relevant part of log here https://gist.github.com/dsyme/054d5567442d3cd3b7e459a5672bb50f

**Comment by dsyme** [comment]
There is one remaining error here:
```
+++ Conformance\ObjectOrientedTypeDefinitions\ClassTypes\AsDeclarations (SanityCheck01.fs) +++
Failed to Find Any Target:  
FAIL
```

I don't understand the failure - it may be a hesien-error related to running tests in parallel.  The SanityCheck01.fs compiles and runs correctly when runn by hand

**Comment by dsyme** [comment]
closed/reopened to trigger CI

**Comment by dsyme** [comment]
OK, all passing now.  I think this RFC/PR is getting quite close to being ready.  I'l lbe doing more testing next week.

**Comment by dsyme** [comment]
I've removed the WIP from this as I believe it is ready to go.

If anyone is able to review this change I'd appreciate it.  Unfortunately it hits the infuriating github limit for diff size, so you will need to review most of the changes in a local git client.

File: vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs [hunk 2]
```diff
@@ -302,7 +283,10 @@ See also ...\SetupAuthoring\FSharp\Registry\FSProjSys_Registration.wxs, e.g.
 
                 this.RegisterProjectFactory(new FSharpProjectFactory(this))
                 //this.RegisterProjectFactory(new FSharpWPFProjectFactory(this :> IServiceProvider))
+<<<<<<< HEAD
```
Comment (ReedCopsey) [inline vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: This looks like a merge artifact...

File: vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs [hunk 3]
```diff
@@ -379,22 +363,36 @@ See also ...\SetupAuthoring\FSharp\Registry\FSProjSys_Registration.wxs, e.g.
                 (project :> ProjectNode)
 
 
+<<<<<<< HEAD
```
Comment (ReedCopsey) [inline vsintegration/src/FSharp.ProjectSystem.FSharp/Project.fs]: And here...

**Comment by ReedCopsey** [comment]
Looks like there are some merge problems in the current code - commented on a couple, but saw quite a few more...

**Comment by dsyme** [comment]
There's a timeout in the AppVeyor testing

**Pull request merged**